paxos/proposer.cc

```diff
@@ -1,125 +1,150 @@
-          static_cast<uint8_t>(i), accept_request, &accept_response);
+      paxos_rpc::PrepareResponse response;
+      grpc::Status status = this->paxos_node_->PrepareGrpc(
+          static_cast<uint8_t>( i ), request, &response);
       if (!status.ok()) {
         LOG(WARNING) << "NODE: [" << static_cast<uint32_t>(node_id_)
-                     << "] Accept grpc failed for node: " << i
+                     << "] Prepare grpc failed for node: " << i
                      << " with error code: " << status.error_code()
                      << " and error message: " << status.error_message();
         continue;
       }
-
-      peer_unchosen_idx[i] = accept_response.first_unchosen_index();
-      if (nop_paxos_round) {
-        continue;
+      if (response.max_chosen_index() > request.index()) {
+        LOG(INFO) << "NODE: [" << static_cast<uint32_t>(node_id_)
+                  << "] Received max chosen index: " << response.max_chosen_index()
+                  << " for index: " << request.index();
+        is_prepare_needed_[i] = true;
+      }
+      else {
+        is_prepare_needed_[i] = false;
       }
 
-      if (accept_response.min_proposal() > request.proposal_number()) {
-        this->replicated_log_->UpdateProposalNumber(
-            accept_response.min_proposal());
-        accept_majority_count = 0;
+      if (response.min_proposal() > request.proposal_number()) {
+        this->replicated_log_->UpdateProposalNumber(response.min_proposal());
+        num_promises = 0;
+        LOG(INFO) << "NODE: [" << static_cast<uint32_t>(node_id_)
+                  << "] Saw a proposal number larger than what we sent, "
+                     "retry Propose operation with a bigger proposal.";
         break;
       }
-      ++accept_majority_count;
-      LOG(INFO) << "NODE: [" << static_cast<uint32_t>(node_id_)
-                << "] Got accept from " << i
-                << " for proposal number: " << accept_response.min_proposal()
-                << " and accepted value: " << value_for_accept_phase;
+      if (response.has_accepted_value()) {
+        if (max_proposal_id < response.accepted_proposal()) {
+          max_proposal_id = response.accepted_proposal();
+          value_for_accept_phase = response.accepted_value();
+          LOG(INFO) << "NODE: [" << static_cast<uint32_t>(node_id_)
+                    << "] Current index already has value: "
+                    << value_for_accept_phase;
+        }
+      }
+      ++num_promises;
     }
+  } while (num_promises < majority_threshold_);
+
+}
 
-    // Since a quorum of acceptors responded with an accept for this value,
-    // we can mark this entry as chosen.
-    if (accept_majority_count >= majority_threshold_) {
-      CHECK(!nop_paxos_round)
-          << "NODE: [" << static_cast<uint32_t>(node_id_)
-          << "] NOP round of paxos should never make log commits.";
+void Proposer::AcceptPhase(paxos_rpc::PrepareRequest& request,
```

Comment:
I don't want to block this review but done should be returned by this function. Instead of a taking it by a reference

Reply:
Can be done as a follow up or something also

Reply:
fixed